### PR TITLE
[DIPU] New implementation for AsyncResourcePoolImpl providing a separate queue for each stream

### PR DESCRIPTION
**Pull request by jfxu-st**
- Implement a partial template specialization for AsyncResourcePoolImpl (algorithm = OneStreamOneQueueAlgo = 1) as a new implementation which provides a separate queue for each stream in the asynchronous resource pool. According to https://github.com/pytorch/pytorch/blob/03a05e791aabe4f2e3938626bff849641dc101ea/c10/cuda/CUDACachingAllocator.cpp#L2745 this implementation can avoid head-of-line delays if one or more streams has long-running operations to wait for.
  - Use the new implementation in BF, BS, RAW caching allocators to replace the old implementation, which has one global queue for all streams.
  - Compared to the old implementation where any event won't be returned to the event pool until all events attached to the same resource are complete, another advantage of the new implementation is that as long as one event is complete, it will be returned to the event pool immediately.
- Remove the restore operation in tensor destruction.
- Remove useless legacy code in DIPUCachingAllocator.cpp
- Fix DIPUEvent's move constructor and move assignment operator by setting `event_` to nullptr inside. If we don't do this, when the original DIPUEvent object (A) is destructed after being moved to a new object (B), the device event `event_` that is still being used by B will be returned to the EventPool, and may be reused in the future while B has no idea about it.

When testing the Llama-2-70B sft task on 64 accelerator cards with tp8 + pp8, which essentially requires two streams (default + communication), compared to old implementation, no performance degradation is observed for this new implementation.

### CODE REVIEW

File: dipu/torch_dipu/csrc_dipu/runtime/core/allocator/DIPUAsyncResourcePool.h [hunk 1]
```diff
@@ -69,4 +73,121 @@ class AsyncResourcePoolImpl : public AsyncResourcePool<T> {
...
+template <class T, at::DeviceType device_type>
+class AsyncResourcePoolImpl<T, device_type, 1> : public AsyncResourcePool<T> {
+ private:
+  // Resources that have events to wait for
+  struct Resource final {
+    const T t;
+    int event_count;
+    const std::deque<DIPUEvent> events;
+
+    Resource(const T& t, int event_count, std::deque<DIPUEvent>& events)
```
Comment (zhaoguochun1995) [inline dipu/torch_dipu/csrc_dipu/runtime/core/allocator/DIPUAsyncResourcePool.h:86]: event_cout 和events.size()是什么关系？

File: dipu/torch_dipu/csrc_dipu/runtime/core/allocator/DIPUAsyncResourcePool.h [hunk 2]
```diff
@@ -69,4 +73,121 @@ class AsyncResourcePoolImpl : public AsyncResourcePool<T> {
...
+  void add(const T& t, std::deque<DIPUEvent>& events) override {
+    std::lock_guard<mutex_t> lk(mutex);
+    if (events.empty()) {
+      queue_without_events.push(t);
+    } else {
+      auto resource = new Resource(t, events.size(), events);
+      for (int i = 0; i < resource->event_count; ++i) {
+        queues_with_events[resource->events[i].stream_id()].emplace(resource,
+                                                                    i);
+      }
```
Comment (zhaoguochun1995) [inline dipu/torch_dipu/csrc_dipu/runtime/core/allocator/DIPUAsyncResourcePool.h:119]: emplace(resource, stream_id) ?  把i 插进去是做什么？

File: dipu/torch_dipu/csrc_dipu/runtime/core/allocator/DIPUAsyncResourcePool.h [hunk 3]
```diff
@@ -69,4 +73,121 @@ class AsyncResourcePoolImpl : public AsyncResourcePool<T> {
...
+
+    if (ready_resource) {
+      return true;
+    }
+
+    for (auto it = queues_with_events.begin();
+         it != queues_with_events.end();) {
+      auto& [stream_id, queue] = *it;
+      auto& [resource, event_id] = queue.front();
+      if (resource->events[event_id].query()) {
```
Comment (zhaoguochun1995) [inline dipu/torch_dipu/csrc_dipu/runtime/core/allocator/DIPUAsyncResourcePool.h:165]: event 如果ready 了，可以提前pop? 通过判断events的size来是否所有event已经ready? event 如果ready 了，就放回event_pool的话，能提升event利用率，同时已经ready的event不会在event_count > 0时 再去判断一遍是否已经ready . ready 这个函数已经不是const的了。

Comment (zhaoguochun1995) [reply dipu/torch_dipu/csrc_dipu/runtime/core/allocator/DIPUAsyncResourcePool.h:165]: 这样在event 比较多时可能能提升一点性能

**Review by zhaoguochun1995 — APPROVED** [review]

File: dipu/torch_dipu/csrc_dipu/runtime/core/DIPUEvent.h [hunk 4]
```diff
@@ -38,8 +38,18 @@ class DIPU_API DIPUEvent {
   DIPUEvent(const DIPUEvent&) = delete;
   DIPUEvent& operator=(const DIPUEvent&) = delete;
 
-  DIPUEvent(DIPUEvent&& other) noexcept = default;
-  DIPUEvent& operator=(DIPUEvent&& other) noexcept = default;
+  DIPUEvent(DIPUEvent&& other) noexcept { *this = std::move(other); }
+
+  DIPUEvent& operator=(DIPUEvent&& other) noexcept {
+    if (this != &other) {
```
Comment (fandaoyi) [inline dipu/torch_dipu/csrc_dipu/runtime/core/DIPUEvent.h:44]: stream_id 也重置下吧.  另外 flag似乎没用了,直接删了吧

File: dipu/torch_dipu/csrc_dipu/runtime/core/allocator/DIPUBFCachingAllocator.cpp [hunk 5]
```diff
@@ -612,8 +611,8 @@ static void deleteBFContext(void* ptr) {
 
 // TODO(allocator) - Refactor it!
 // NOLINTBEGIN(cppcoreguidelines-avoid-non-const-global-variables,modernize-avoid-bind)
-DIPU_REGISTER_ALLOCATOR(BF, DIPU_DEVICE_TYPE_MACRO, BFCachingAllocator, 0);
-DIPU_REGISTER_ALLOCATOR(BF, CPU, BFCachingAllocator, 0);
+DIPU_REGISTER_ALLOCATOR(BF, DIPU_DEVICE_TYPE_MACRO, BFCachingAllocator, 1, 0);
+DIPU_REGISTER_ALLOCATOR(BF, CPU, BFCachingAllocator, 1, 0);
```
Comment (fandaoyi) [inline dipu/torch_dipu/csrc_dipu/runtime/core/allocator/DIPUBFCachingAllocator.cpp:615]: 1 容易迷惑, 给个有意义的常量吧. 比如 XXRssAlg 啥的.

**Review by mrdanielw — APPROVED** [review]

**Pull request merged**